internal/pkg/graph/graph.go

```diff
@@ -0,0 +1,97 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package graph provides functionality for directed graphs.
+package graph
+
+// nodeColor denotes the color of a node when running DFS in a graph.
+type nodeColor int
+
+const (
+	white nodeColor = iota + 1
+	gray
+	black
+)
+
+// Graph represents a directed graph.
+type Graph struct {
+	Nodes map[string][]string
+}
+
+// New initiates a new Graph.
+func New() *Graph {
+	return &Graph{
+		Nodes: make(map[string][]string),
+	}
+}
+
+// Add adds a connection between two Nodes.
+func (g *Graph) Add(fromNode, toNode string) {
```

Comment:
nit: Just something to consider: https://dave.cheney.net/2019/09/24/be-wary-of-functions-which-take-several-parameters-of-the-same-type 

When reading the code:
```go
g.Add("efe", "penghao") // it's not clear which one is the from and to 
```
Maybe, an alternative:
```go
type Edge struct {
  From string
  To string
}
func (g *Graph) Add(edge Edge) { ... }
```